### ISSUE TITLE
[BUG][PSA: DATA LOSS] Arborist loses all data upon saving/changing scenes

### ISSUE THREAD

**Issue by aXu-AP** [dreadpon/godot_spatial_gardener#48]
**Godot version**
v4.2.2.stable.official [15073afe3]

**Plugin version**
1.3.2 (asset library)

**Platform**
Windows 10

**Issue description**
Okay this has happened a few times already, but I'm not that far into the project that catastrophe has happened. Some times when I save another scene and change to the scene with gardener, the plugin simply goes haywire and stops working, spewing endless stream of errors:
![kuva](https://github.com/dreadpon/godot_spatial_gardener/assets/1621768/1a7f66af-d43b-4d11-a15a-33ebed18edf0)
From this situation I've tried 2 different options:
1. Save and reload the project. After reloading I'm greeted with following errors:
```
res://addons/dreadpon.spatial_gardener/arborist/arborist.gd:89 - Invalid call. Nonexistent function 'duplicate_tree' in base 'Nil'.
INFO: gardener.gd (Gardener): Found existing Arborist
  res://addons/dreadpon.spatial_gardener/arborist/mmi_octree/mmi_octree_node.gd:316 - Invalid set index 'cast_shadow' (on base: 'Nil') with value of type 'int'.
  res://addons/dreadpon.spatial_gardener/arborist/mmi_octree/mmi_octree_node.gd:316 - Invalid set index 'cast_shadow' (on base: 'Nil') with value of type 'int'.
  res://addons/dreadpon.spatial_gardener/arborist/mmi_octree/mmi_octree_node.gd:316 - Invalid set index 'cast_shadow' (on base: 'Nil') with value of type 'int'.
  res://addons/dreadpon.spatial_gardener/arborist/mmi_octree/mmi_octree_node.gd:316 - Invalid set index 'cast_shadow' (on base: 'Nil') with value of type 'int'.
  res://addons/dreadpon.spatial_gardener/arborist/mmi_octree/mmi_octree_node.gd:526 - Cannot call method 'get_node_or_null' on a null value.
  res://addons/dreadpon.spatial_gardener/arborist/mmi_octree/mmi_octree_node.gd:567 - Cannot call method 'get_child_count' on a null value.
  res://addons/dreadpon.spatial_gardener/arborist/mmi_octree/mmi_octree_node.gd:526 - Cannot call method 'get_node_or_null' on a null value.
  res://addons/dreadpon.spatial_gardener/arborist/mmi_octree/mmi_octree_node.gd:567 - Cannot call method 'get_child_count' on a null value.
  res://addons/dreadpon.spatial_gardener/arborist/mmi_octree/mmi_octree_node.gd:526 - Cannot call method 'get_node_or_null' on a null value.
  res://addons/dreadpon.spatial_gardener/arborist/mmi_octree/mmi_octree_node.gd:567 - Cannot call method 'get_child_count' on a null value.
  res://addons/dreadpon.spatial_gardener/arborist/mmi_octree/mmi_octree_node.gd:526 - Cannot call method 'get_node_or_null' on a null value.
  res://addons/dreadpon.spatial_gardener/arborist/mmi_octree/mmi_octree_node.gd:567 - Cannot call method 'get_child_count' on a null value.
  res://addons/dreadpon.spatial_gardener/arborist/mmi_octree/mmi_octree_node.gd:526 - Cannot call method 'get_node_or_null' on a null value.
  res://addons/dreadpon.spatial_gardener/arborist/mmi_octree/mmi_octree_node.gd:567 - Cannot call method 'get_child_count' on a null value.
  res://addons/dreadpon.spatial_gardener/arborist/mmi_octree/mmi_octree_node.gd:526 - Cannot call method 'get_node_or_null' on a null value.
  res://addons/dreadpon.spatial_gardener/arborist/mmi_octree/mmi_octree_node.gd:567 - Cannot call method 'get_child_count' on a null value.
  res://addons/dreadpon.spatial_gardener/arborist/mmi_octree/mmi_octree_node.gd:526 - Cannot call method 'get_node_or_null' on a null value.
  res://addons/dreadpon.spatial_gardener/arborist/mmi_octree/mmi_octree_node.gd:567 - Cannot call method 'get_child_count' on a null value.
  res://addons/dreadpon.spatial_gardener/arborist/mmi_octree/mmi_octree_node.gd:526 - Cannot call method 'get_node_or_null' on a null value.
  res://addons/dreadpon.spatial_gardener/arborist/mmi_octree/mmi_octree_node.gd:567 - Cannot call method 'get_child_count' on a null value.
  res://addons/dreadpon.spatial_gardener/arborist/mmi_octree/mmi_octree_node.gd:316 - Invalid set index 'cast_shadow' (on base: 'Nil') with value of type 'int'.
```
All the props are in place, but the gardener doesn't recognize them. Inspecting diff of the scene file it seems that some data has been permanently lost.

**OR**

2. Reload saved scene. This seems to fix the issue, so as long as I know to do this I can prevent data loss (unless there was unsaved changes).

And this is where the PSA comes in: 
### If Gardener starts spamming errors, reload saved scene. DO NOT SAVE.

**Steps to Reproduce**
1. Make a scene with Gardener (let's say `test.tscn`), and second unrelated scene (`character.tscn`) instantiated in that scene.
2. Have both scenes opened in editor.
3. Save the `character.tscn`.
4. Change tabs to `test.tscn`.

**Minimal reproduction project**
[GardererMRP.zip](https://github.com/dreadpon/godot_spatial_gardener/files/15193308/GardererMRP.zip)
Do steps 2-4 with this MRP.

**Comment by dreadpon** [comment]
Might be a problem with destruction code added in 1.3.2 which executes in _exit_tree, which apparently is called when switching scenes.

Will need to work this version and if true, look for an alternative which triggers on the actual node deletion (when scene is being closed), probably a NOTIFICATION_PREDELETE

**Comment by dreadpon** [comment]
If you have the time to test the MRP in 1.3.1 - this might help in locating the problem

**Comment by aXu-AP** [comment]
I can confirm that this is indeed a regression in 1.3.2. I couldn't reproduce the bug with 1.3.1

**Comment by dreadpon** [comment]
Fixed with https://github.com/dreadpon/godot_spatial_gardener/commit/902fe9c1e4de209b2d9d867e297335c14cc40e91

But not 100% sure some echo of this problem will not return in the future. The fix was weird. Turns out we _were_ already using `NOTIFICATION_PREDELETE`, but inside it we set certain references to `null` and it really messed with Godot's resource caching (?) system. These certain references weren't mandatory to be set to `null`, and it's probably better if they aren't, but the observed problem and the fix don't quite add up, especially considering the trigger requirements.

Will count this as closed for now, we'll return to this if further problems are revealed on your side @aXu-AP. No release yet, please use master. Found other things I need to investigate and fix before releasing 1.3.3.

Will keep this Issue open for PSA visibility.

**Comment by dreadpon** [comment]
The fix above seems to cause bugs in managing MultiMeshes (some of them get left behind, leading to weird behavior). Discovered by running autotest on 1.3.2 compared to 1.3.1.

Fixed in 729b413 by reverting some of the changes (the fix itself still applies).

**Comment by dreadpon** [comment]
@aXu-AP would appreciate it if you put the `master` version through some basic real-world work cases to see if it's stable enough to release a hotfix. Your MRP seems to work fine though, and my autotests didn't show any more problems.

**Comment by aXu-AP** [comment]
I updated Gardener to master branch in my game, and so far it does no problems have risen 👍

**Comment by dreadpon** [comment]
Good to know, will make a release tomorrow.
Don't forget to use Git!)

**Comment by aXu-AP** [comment]
Yeah I did do a commit before in case something goes wrong 😎

**Comment by dreadpon** [comment]
Closed by 902fe9c1e4de209b2d9d867e297335c14cc40e91
